### ISSUE TITLE
Bug: array-element-newline with multiline: true

### ISSUE THREAD

**Issue by MatthewHerbst** [eslint/eslint#8728]
**Tell us about your environment**

* **ESLint Version:** `4.0.0`
* **Node Version:** `6.5.0`
* **npm Version:** `3.10.3`

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**
My whole config is massive because I have most rules listed. Here is the relevant part that was added.

```
'array-element-newline': [2, { multiline: true }]
```
You can find the entire file (sans the new rule) [here](https://github.com/brigade/eslint-config-brigade/blob/v5.0.0/index.js).

**What did you do? Please include the actual source code causing the issue.**
This code exists in the file currently:

```js
'import/order': [2, {
  groups: [
    'builtin',
    'external',
    'internal',
    ['index', 'sibling', 'parent'],
  ],
}],
```

The new rule ([`array-element-newline`](http://eslint.org/docs/rules/array-element-newline)) properly yelled at me that the `2` should be on its own line. So I fixed that, and now have:

```js
'import/order': [
  2,
  {
    groups: [
      'builtin',
      'external',
      'internal',
      ['index', 'sibling', 'parent'],
    ],
  }
],
```

**What did you expect to happen?**
I expect this to be valid:

```js
'import/order': [
  2,
  {
    groups: [
      'builtin',
      'external',
      'internal',
      ['index', 'sibling', 'parent'],
    ],
  }
],
```

**What actually happened? Please include the actual, raw output from ESLint.**

The new rule is complaining for the items within the `groups` array that:

> There must be no linebreak here. (array-element-newline)

Playing around with it a little bit, it seems to be getting confused by the last item, which is itself a list. Changing the code to this fixes it:

```js
'import/order': [
  2,
  {
    groups: [
      'builtin',
      'external',
      'internal',
      [
        'index', 'sibling', 'parent'],
    ],
  }
],
```
Clearly that is not what this rule intends.

----------
I'd be happy to work on a fix for this, but will be leaving the country in a few days and not returning until the 26th. Reporting this now because that seems like a long time to wait potentially.

**Comment by not-an-aardvark** [comment]
Thanks for reporting; however, this is the intended behavior. The [`multiline`](http://eslint.org/docs/rules/array-element-newline#options) option for `array-element-newline` indicates that newlines between elements are required whenever there is any element that spans multiple lines.

* In the outer array in your example, the object with a `groups` key spans multiple lines, so it needs to be separated from the `2` by a newline.
* However, in the `groups` array, there are at first no elements that span multiple lines, so the rule forbids linebreaks between the elements.
* After you add a linebreak, the last element of the `groups` array spans multiple lines, so linebreaks are required between the elements.

If you'd like to allow linebreaks in the `groups` array even though none of its elements span multiple lines, one option would be to use the `multiline` option in conjunction with the `minItems` option (to specify that arrays with at least 4 elements also need linebreaks between the elements).

**Comment by MatthewHerbst** [comment]
@not-an-aardvark thanks for the response, but I think you lost be a little bit. Just to confirm:

This is supposed to pass (because it does)?

```
'import/order': [
  2,
  {
    groups: [
      'builtin',
      'external',
      'internal',
      [
        'index', 'sibling', 'parent'],
    ],
  }
],
```

I'm not a style expert, but nothing about that seems right to me, and if that is desired, I question how it ever made it into this library...

**Comment by not-an-aardvark** [comment]
In that example, the `groups` array has this element:

```js
[
  'index', 'sibling', 'parent']
```

That element spans multiple lines. Since you've configured the the `array-element-newline` rule with the `multiline: true` option, it will enforce newlines between the elements of the `groups` array since that array contains an element spanning multiple lines.

> I'm not a style expert, but nothing about that seems right to me, and if that is desired, I question how it ever made it into this library...

The rule is doing what it says on the tin for the way you've configured it. If you'd like to disallow code like this, you might be interested in enabling the [`array-bracket-newline`](http://eslint.org/docs/rules/array-bracket-newline) rule to prevent the inner array from being formatted like that.

**Comment by MatthewHerbst** [comment]
Appreciate the quick response, sorry if I'm taking too much time on this.

Is there any way to configure the rules to enforce what I had expected them to?

```
'import/order': [
  2,
  {
    groups: [
      'builtin',
      'external',
      'internal',
      ['index', 'sibling', 'parent'],
    ],
  }
],
```

I guess what I'm looking for is:

1. allow a single-line array to have unlimited items
2. any multi-line array must have each element on a new line
3. complex array members conform the rules above

I don't think using `minItems` (with either) rule will help me here because I can't make any promise across our code base about the number of items in a a single-line or multi-line array.

**Comment by platinumazure** [comment]
It seems like the ask is to have a multiline setting where

* If an element takes multiple lines, then each consecutive pair of elements MUST have a newline between them;
* If no element takes multiple lines, then each consecutive pair of elements MAY have a newline between them;
* If any elements have a newline between them, all elements must have a newline between them for consistency.

Am I thinking in the right direction?

With the current configuration, this is what ESLint expects:

```js
'import/order': [
  2,
  {
    groups: ['builtin', 'external', 'internal', ['index', 'sibling', 'parent']],
  }
],
```

This is because none of the elements in the `groups` array is multiline.

**Comment by MatthewHerbst** [comment]
@platinumazure that seems pretty accurate. I guess when I implemented this new rule, I wasn't expecting it to prevent me from doing the second bullet point; I only expected it to enforce the first and third points.

In general, I think there's a big difference between a rule enforcing particular behavior and a rule banning particular behavior, and I think most rules should attempt to do one or the other, not both.

**Comment by not-an-aardvark** [comment]
> In general, I think there's a big difference between a rule enforcing particular behavior and a rule banning particular behavior, and I think most rules should attempt to do one or the other, not both.

I'm confused by this -- could you clarify? It seems like enforcing a particular behavior would necessarily involve banning all other behaviors.

**Comment by MatthewHerbst** [comment]
Sure, let's use the [`array-element-newline`](http://eslint.org/docs/rules/array-element-newline) rule as an example. I'm going to go off the documentation in the rule. If the documentation is misleading or incomplete, that's a different issue I guess, but for now I'm going to use it as the source of truth for what the rule is meant to do:

> A number of style guides require or disallow line breaks between array elements.
>
> This rule enforces line breaks between array elements.
> ...
> `"multiline": <boolean>` requires line breaks if there are line breaks inside elements. If this is false, this condition is disabled.

The rule clearly documents @platinumazure's first and third bullet points:

> If any elements have a newline between them, all elements MUST have a newline between them for consistency.

> If an element takes multiple lines, then each consecutive pair of elements MUST have a newline between them;

However, at no point does it say or hint at (including in the example pass/fail code in the rule docs) that you _may only_ have line break between elements _if, and only if_ one or more of those elements are themselves multiline.

Basically, there is nothing in that rule to suggest that it should stop me from doing (@platinumazure's second bullet point):

```
'import/order': [
  2,
  {
    groups: [
      'builtin',
      'external',
      'internal',
      ['index', 'sibling', 'parent'],
    ],
  }
],
```

To get simpler, even this example is not allowed:

```
plugins: [
  'import',
],
```

I would argue that both of the above perfectly comply with the rule _as the docs are written_. At no point does the rule suggest that you may only have newlines if-and-only-if something spans multiple lines. Rather, it simply says that you must have newlines if something spans multiple lines. That is a  subtle but important difference.

**Comment by MatthewHerbst** [comment]
@platinumazure ah, not sure why I didn't pick up on that part when reading through again. Thanks for highlighting that.

I would be in favor of such an enhancement. Unless someone picks it up, I'm happy to work on a PR for it when I get back if that's something this library would be in favor of.

**Comment by evilebottnawi** [comment]
Seems same as `consistent` in http://eslint.org/docs/rules/object-curly-newline

**Comment by not-an-aardvark** [comment]
Closing because this is working as intended, but feel free to create another issue or make a PR if you'd like to propose a new option for the rule.
